jailer.go

```diff
@@ -0,0 +1,388 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package firecracker
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+)
+
+const (
+	defaultJailerPath = "/srv/jailer/firecracker"
+	defaultJailerBin  = "jailer"
+
+	rootfsFolderName = "root"
+)
+
+// SeccompLevelValue represents a secure computing level type.
+type SeccompLevelValue int
+
+// secure computing levels
+const (
+	// SeccompLevelDisable is the default value.
+	SeccompLevelDisable SeccompLevelValue = iota
+	// SeccompLevelBasic prohibits syscalls not whitelisted by Firecracker.
+	SeccompLevelBasic
+	// SeccompLevelAdvanced adds further checks on some of the parameters of the
+	// allowed syscalls.
+	SeccompLevelAdvanced
+)
+
+// JailerConfig is jailer specific configuration needed to execute the jailer.
+type JailerConfig struct {
+	// GID the jailer switches to as it execs the target binary.
+	GID *int
+
+	// UID the jailer switches to as it execs the target binary.
+	UID *int
+
+	// ID is the unique VM identification string, which may contain alphanumeric
+	// characters and hyphens. The maximum id length is currently 64 characters
+	ID string
+
+	// NumaNode represents the NUMA node the process gets assigned to.
+	NumaNode *int
+
+	// ExecFile is the path to the Firecracker binary that will be exec-ed by
+	// the jailer. The user can provide a path to any binary, but the interaction
+	// with the jailer is mostly Firecracker specific.
+	ExecFile string
+
+	// ChrootBaseDir represents the base folder where chroot jails are built. The
+	// default is /srv/jailer
+	ChrootBaseDir string
+
+	// NetNS represents the path to a network namespace handle. If present, the
+	// jailer will use this to join the associated network namespace
+	NetNS string
+
+	//  Daemonize is set to true, call setsid() and redirect STDIN, STDOUT, and
+	//  STDERR to /dev/null
+	Daemonize bool
+
+	// SeccompLevel specifies whether seccomp filters should be installed and how
+	// restrictive they should be. Possible values are:
+	//
+	//	0 : (default): disabled.
+	//	1 : basic filtering. This prohibits syscalls not whitelisted by Firecracker.
+	//	2 : advanced filtering. This adds further checks on some of the
+	//			parameters of the allowed syscalls.
+	SeccompLevel SeccompLevelValue
+
+	// DevMapperStrategy will dictate how files are transfered to the root drive.
+	DevMapperStrategy HandlersAdaptor
```

Comment:
I think we need a different name for this.  "DevMapper" to me implies [device mapper](https://en.wikipedia.org/wiki/Device_mapper), which this does not use.

How about FileMapperStrategy or ChrootStrategy?